### PR TITLE
feat: working mvp with native BATCH message support

### LINKED ISSUE #1: feat: native Snowflake bulk load for `BATCH` messages
Related:

- SDK `BATCH` docs: https://meltano-sdk--904.org.readthedocs.build/en/904/batch.html
- Snowflake `COPY INTO` docs for bulk load: https://docs.snowflake.com/en/user-guide/data-load-local-file-system.html
- Corresponding feature on `tap-snowflake`: https://github.com/MeltanoLabs/tap-snowflake/issues/1

Spec details:

1. Use table stages to avoid requiring S3 creds.
2. Use default encoding of `jsonl` and `gzip` to match initial built-in SDK feature spec.
3. Skip record validation to avoid needing to parse the raw files.
4. One `COPY` command per `BATCH` message's `manifest`. (Each manifest could contain 1 or many files, depending on the tap implementation.)

### PR DESCRIPTION
**Pull request by kgpayne**
Closes #1 

- #1

### CODE REVIEW

File: target_snowflake/sinks.py [hunk 1]
```diff
@@ -23,10 +33,238 @@ def get_sqlalchemy_url(self, config: dict) -> str:
...
+        if not self.allow_column_alter:
+            raise NotImplementedError(
+                "Altering columns is not supported. "
+                f"Could not convert column '{full_table_name}.{column_name}' "
+                f"from '{current_type}' to '{compatible_sql_type}'."
+            )
+
+        self.connection.execute(
+            sqlalchemy.DDL(
+                "ALTER TABLE %(table)s ALTER COLUMN %(col_name)s SET DATA TYPE %(col_type)s",
```
Comment (kgpayne) [inline target_snowflake/sinks.py:93]: This is the only line that changed in this function compared to super(). Will raise a PR move this `ALTER` statement into a `get` method that can be overridden.

**Review by aaronsteers — APPROVED** [review]

File: target_snowflake/sinks.py [hunk 2]
```diff
@@ -23,10 +61,332 @@ def get_sqlalchemy_url(self, config: dict) -> str:
...
+    # TODO: update once https://github.com/meltano/sdk/pull/1114 merges
+    def _adapt_column_type(
+        self,
+        full_table_name: str,
+        column_name: str,
+        sql_type: sqlalchemy.types.TypeEngine,
+    ) -> None:
+        """Adapt table column type to support the new JSON schema type.
+
+        Args:
```
Comment (aaronsteers) [inline target_snowflake/sinks.py:107]: Tiny nit:

```suggestion
    def _adapt_column_type(
        self,
        full_table_name: str,
        column_name: str,
        sql_type: sqlalchemy.types.TypeEngine,
    ) -> None:
        """Adapt table column type to support the new JSON schema type.

        Overridden here as Snowflake ALTER syntax for columns is different than that implemented in the SDK
        https://docs.snowflake.com/en/sql-reference/sql/alter-table-column.html
        TODO: update once https://github.com/meltano/sdk/pull/1114 merges

        Args:
```

File: target_snowflake/sinks.py (hunk 1 above)
Comment (aaronsteers) [reply target_snowflake/sinks.py:142]: 👍 - Thanks for calling this out. Now it's in the method declaration so I'll resolve this comment.

File: target_snowflake/sinks.py [hunk 3]
```diff
@@ -23,10 +61,332 @@ def get_sqlalchemy_url(self, config: dict) -> str:
...
+            TypeMap(th._jsonschema_type_check, sct.NUMBER(), ("integer",)),
+            TypeMap(th._jsonschema_type_check, sct.VARIANT(), ("object",)),
+            TypeMap(th._jsonschema_type_check, sct.VARIANT(), ("array",)),
+        ]
+        # apply type maps
+        if th._jsonschema_type_check(jsonschema_type, ("string",)):
+            datelike_type = th.get_datelike_property_type(jsonschema_type)
+            target_type = evaluate_typemaps(sting_submaps, datelike_type, target_type)
+        else:
+            target_type = evaluate_typemaps(type_maps, jsonschema_type, target_type)
```
Comment (aaronsteers) [inline target_snowflake/sinks.py:185]: Note to our future selves to see what can be improved upstream in the SDK to simplify and remove need to call the private `_json_schema_type_check()`

Comment (aaronsteers) [reply target_snowflake/sinks.py:185]: To discuss further in:

- #4 

as follow-on.

File: target_snowflake/sinks.py [hunk 4]
```diff
@@ -23,10 +61,332 @@ def get_sqlalchemy_url(self, config: dict) -> str:
...
+        Returns:
+            The SQLAlchemy type representation of the data type.
+        """
+        # start with default implementation
+        target_type = SQLConnector.to_sql_type(jsonschema_type)
+        # snowflake max and default varchar length
+        # https://docs.snowflake.com/en/sql-reference/intro-summary-data-types.html
+        maxlength = jsonschema_type.get("maxLength", 16777216)
+        # define type maps
+        sting_submaps = [
```
Comment (aaronsteers) [inline target_snowflake/sinks.py:169]: Minor typo:

```suggestion
        string_submaps = [
```

And below when referencing the var.

**Comment by aaronsteers** [comment]
@kgpayne - This looks ready to go! But please do go ahead and log any issues that you want to make sure we come back to. Even for SDK upstream changes, you can log an issue here in this repo to get the conversations started - and to represent the work that would be needed after the SDK-side updates.

A huge step from what's on `main` so I would lean towards a merge this week if we can. 🎉 

@edgarrmondragon - Any changes you'd like to see, now, or in an upcoming iteration?

File: target_snowflake/sinks.py [hunk 5]
```diff
@@ -1,8 +1,46 @@
 """Snowflake target sink class, which handles writing streams."""
-
 from __future__ import annotations
 
+import gzip
+import json
+import os
+from gzip import open as gzip_open
```
Comment (edgarrmondragon) [inline target_snowflake/sinks.py:7]: I don't think this is used

```suggestion
```

**Review by edgarrmondragon — CHANGES_REQUESTED** [review]

**Comment by edgarrmondragon** [comment]
@kgpayne Just a nit on an unused import, otherwise looks good.

We should probably add pre-commit to lint the repo.

**Review by aaronsteers — APPROVED** [review]

**Comment by aaronsteers** [comment]
> @kgpayne Just a nit on an unused import, otherwise looks good.
> 
> We should probably add pre-commit to lint the repo.

Logged:

- https://github.com/MeltanoLabs/target-snowflake/issues/5

**Review by edgarrmondragon — APPROVED** [review]

**Comment by aaronsteers** [comment]
I'm going to go ahead and merge this, and we can iterate from here.

cc @pnadolny13, @visch, @BuzzCutNorman

Follow-ups are logged in the issue tracker: https://github.com/meltanolabs/target-snowflake/issues

Nice work on this, @kgpayne

**Pull request merged**